bin/recategorize-messages.py

```diff
@@ -0,0 +1,145 @@
+        .join(Namespace.account)
+        .filter(Account.discriminator.in_(discriminators))
+    )
+    if only_account_id:
+        namespace_query = namespace_query.filter(
+            Namespace.account_id == only_account_id
+        )
+
+    with global_session_scope() as session:
+        return namespace_query.with_session(session).count()
+
+
+def yield_account_id_and_message_ids(
+    *,
+    only_account_id: int | None,
+    date_start: datetime.date | None,
+    date_end: datetime.date | None,
+    only_inbox: bool,
+    only_types: set[AccountType] = ALL_ACCOUNT_TYPES,
+) -> Iterable[int, list[int]]:
+    discriminators = {account_type + "account" for account_type in only_types}
+    namespace_query = (
+        Query([Namespace.account_id, Namespace.id])
+        .join(Namespace.account)
+        .filter(Account.discriminator.in_(discriminators))
+    )
+    if only_account_id:
+        namespace_query = namespace_query.filter(
+            Namespace.account_id == only_account_id
+        )
+
+    with global_session_scope() as session:
+        account_id_to_namespace_id = {
+            account_id: namespace_id
+            for account_id, namespace_id in namespace_query.with_session(session)
+        }
+
+    for account_id, namespace_id in account_id_to_namespace_id.items():
+        query = Query([Message.id]).filter(Message.namespace_id == namespace_id)
+
+        if only_inbox:
+            inbox_folder = ImapUid.folder.has(Folder._canonical_name == "INBOX")
+            query = query.filter(Message.imapuids.any(inbox_folder))
+        if date_start:
+            query = query.filter(Message.created_at >= date_start)
+        if date_end:
+            query = query.filter(Message.created_at < date_end)
+
+        with global_session_scope() as session:
+            message_ids = [message_id for message_id, in query.with_session(session)]
+
+        yield account_id, message_ids
+
+
+@click.command()
+@click.option("--date-start", type=click.DateTime(formats=["%Y-%m-%d"]), default=None)
+@click.option("--date-end", type=click.DateTime(formats=["%Y-%m-%d"]), default=None)
+@click.option("--only-account-id", type=int, default=None)
+@click.option("--only-inbox", is_flag=True, default=False)
+@click.option("--only-types", default=",".join(ALL_ACCOUNT_TYPES))
+@click.option("--dry-run/--no-dry-run", default=True)
+def main(
+    only_account_id: int | None,
+    only_inbox: bool,
+    only_types: str,
+    date_start: datetime.date | None,
+    date_end: datetime.date | None,
+    dry_run: bool,
+) -> None:
+    print(
+        f"Settings: {only_account_id=}, {only_inbox=}, {date_start=}, {date_end=}, {dry_run=}\n"
+    )
+
+    total_namespace_count = get_total_namespace_count(
+        only_account_id=only_account_id, only_types=set(only_types.split(","))
+    )
+    print(f"{total_namespace_count=}\n")
+
+    def session_factory():
+        return global_session_scope() if dry_run else session_scope(None)
```

Comment:
This is sync-engine curiosity. sync-engine has two ways of managing transactions... `global_session_scope` is for reads and `session_scope` is mainly meant for writes. `session_scope` will automatically commit at the end. And it also automatically behind the scenes serializes changes to transaction log, and that transaction log is what CRM consumes.